### DISCUSSION TITLE
How do I use StyledComponents in the Next.js 13 app router?

### DISCUSSION THREAD

**Discussion by brun0xff** [vercel/next.js#53177] [App Router]
From the [Next.js docs](https://beta.nextjs.org/docs/styling/css-in-js):

>Warning: CSS-in-JS libraries that require runtime JavaScript are not currently supported in Server Components.
>
>The following libraries are supported in Client Components in the app directory:
>
>styled-jsx
>styled-components
>style9

**Questions**

1 - Does that mean I have to use the use-client directive in every component that uses StyledComponents? If that's the case, it makes sense to use Styled Components if Server Components are not feasible.

2 - If so, is it possible to write all StyledComponents in one file, with the use-client directive at the top, and then just import these Client Components as needed into the Server Components?

3 - Will the approach above somehow turn the Server Components into Client Components?

4 - The Next.js docs state that all components within the app directory are Server Components by default. If we save a component outside the app directory and import it into a Server Component in the app directory, will that imported function then also become a Server Component?

**Comment by icyJoseph** [comment]
> 1 - Does that mean I have to use the use-client directive in every component that uses StyledComponents? If that's the case, it makes sense to use Styled Components if Server Components are not feasible.

I think so yeah, since there needs to be a runtime/client-side evaluation of the styles. However, do remember that this means business as usual. Client Components are the components we've used up till now. The new thing are Server Components.

> 2 - If so, is it possible to write all StyledComponents in one file, with the use-client directive at the top, and then just import these Client Components as needed into the Server Components?

Yes. That'll make this module a client side boundary. Not directly related to Styled Components, but here for example I made these icons client components, https://github.com/icyJoseph/icyJoseph/blob/dev/src/design-system/Icons.tsx  

> 3 - Will the approach above somehow turn the Server Components into Client Components?

This is where it gets a bit tricky, but I think the short answer is no. If a module has `use client` at the top, it creates a client component tree, which could potentially be weaved/slotted-in to a Server Component, but still be a client component.

I do believe the other way around is a far more common issue. Because a component module without 'use client', gets opted into Server Components, unless it is part of a 'use client' module tree, in which case it is turned into a client component. 

Imagine an async component that reads from a db, and ends up imported into a file with `use client`, that'll break. However, importing a 'use client' component into a page that's behaving as a server component, is the actual use case we want!

> 4 - The Next.js docs state that all components within the app directory are Server Components by default. If we save a component outside the app directory and import it into a Server Component in the app directory, will that imported function then also become a Server Component?

I think you are reading too tightly into the wording there. What the docs mean is that the App Dir Component tree, with its imports, is by default, aggressively opted into Server Components. If there's no use client directive at all in the entire module import tree, then that entire app is a Server Component. 

I think you should read these:

- https://demystifying-rsc.vercel.app/
- https://github.com/reactwg/server-components/discussions/4

**Comment by silas-v1** [comment]
[Following Next.js recommendation for configuration](https://nextjs.org/docs/app/building-your-application/styling/css-in-js#styled-components), each styled component accesses the context provided by the styled component registry; therefore the components must be client-rendered. If you attempt to render a styled-component without 'use client' directive, Next.js will throw a context error like: 
```
TypeError: createContext only works in Client Components. Add the "use client" directive at the top of the file to use it. Read more: https://nextjs.org/docs/messages/context-in-server-component
```
